Handle incompatible IID SDK version with FIS SDK.

ankitaj224 opened the pull request.

andirayo commented on firebase-installations/firebase-installations.gradle:
```diff
@@ -47,6 +47,10 @@ dependencies {
+    compileOnly ("com.google.firebase:firebase-iid:20.0.2"){
+        exclude group: "com.google.firebase", module: "firebase-common"
+        exclude group: "com.google.firebase", module: "firebase-components"
+    }
```
I don't understand this.
Can you please explain to me in person?

andirayo replied:
open question

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -73,6 +78,11 @@
   /* file used for process-level syncronization of generating and persisting fids */
   private static final String LOCKFILE_NAME_GENERATE_FID = "generatefid.lock";
 
+  private static final String IID_SDK_PROP_FILE = "/firebase-iid.properties";
```
Please add a code comment why you need this.
Something like
// Values that are needed to prevent incompatible versions of Instance-ID SDK to be used in parallel

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -73,6 +78,11 @@
   private static final String LOCKFILE_NAME_GENERATE_FID = "generatefid.lock";
 
+  private static final String IID_SDK_PROP_FILE = "/firebase-iid.properties";
+  private static final String IID_SDK_VERSION_BEFORE_FIS = "20.0.2";
```
why only this explicit version?
is this the version of IID since I/O 2019?

ankitaj224 replied:
This is the latest version of IID with no on dependency on FIS SDK.

https://mvnrepository.com/artifact/com.google.firebase/firebase-iid

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -133,8 +143,17 @@ public static FirebaseInstallations getInstance() {
+          "FirebaseInstallations will not work correctly with current version of Firebase "
+              + "Instance ID. Please update your Firebase Instance ID version.",
+          Status.INCOMPATIBLE_VERSIONS);
+    }
```
Please outsource in dedicated method!

#preventParallelUsageOfIncompatibleVersionOfInstanceIdSDk

andirayo replied:
let the other method throw the exception!

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -451,4 +470,34 @@ private Void deleteFirebaseInstallationId() throws FirebaseInstallationsExceptio
+      return true;
+    }
+    return true;
+  }
```
Please have this method throw an Exception

ankitaj224 replied:
Can't. I need to handle the NoClassDefFoundError exception and return true. Its the best case when there is no IID SDK in gradle depedency.

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -451,4 +470,34 @@ private Void deleteFirebaseInstallationId() throws FirebaseInstallationsExceptio
+      }
+    } catch (NoClassDefFoundError exception) {
+      // There is no IID SDK at all.
+      return true;
```
code duplication of `return true`

ankitaj224 replied:
Updated the code better readability. Please review again.

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -451,4 +470,34 @@ private Void deleteFirebaseInstallationId() throws FirebaseInstallationsExceptio
+    return true;
+  }
+
+  private static String getIIDVersion() {
```
@Nullable

andirayo replied:
How about using Optional?

ankitaj224 replied:
Added the annotation. Can't add Optional - min SDK requirement for Optional is 24 and firebase sdk's are at 14

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -451,4 +470,34 @@ private Void deleteFirebaseInstallationId() throws FirebaseInstallationsExceptio
+        props.load(fs);
+        return props.getProperty("version", /* defaultValue */ null);
+      }
+    } catch (IOException e) {
```
is this the only exception that can be thrown here?

ankitaj224 replied:
Yes, its the only expected exception.

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/Utils.java:
```diff
@@ -47,4 +47,28 @@ public boolean isAuthTokenExpired(PersistedInstallationEntry entry) {
+  /**
+   * Compares two version strings.
+   *
+   * @note It assumes version is in the "1.0.1" format.
```
I don't like it

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/Utils.java:
```diff
@@ -47,4 +47,28 @@ public boolean isAuthTokenExpired(PersistedInstallationEntry entry) {
+        return comparison;
+      }
+    }
+    return Integer.compare(version1Parts.length, version2Parts.length);
```
This seems way too complicated to me...
How about we just check for the two last versions...
or all versions since last I/O and disallow those?

ankitaj224 replied:
Done.

vkryachko commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -451,4 +477,35 @@ private Void deleteFirebaseInstallationId() throws FirebaseInstallationsExceptio
+
+  private static boolean isIIDVersionCompatible() {
+    try {
+      final Class<FirebaseInstanceId> firebaseInstanceIdClass = FirebaseInstanceId.class;
```
Proguard may end up removing this line due to the `code/removal/variable` optimization, I would suggest you test with this optimization as well as with `code/removal/advanced`, `code/removal/simple` optimizations.

andirayo replied:
+1

ankitaj224 replied:
Done. PTAL.

vkryachko replied:
Could you explain how it was done? Can we have an integ test for this?

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -121,7 +135,7 @@ private void preConditionChecks() {
    */
   @NonNull
-  public static FirebaseInstallations getInstance() {
+  public static FirebaseInstallations getInstance() throws FirebaseInstallationsException {
```
Using an incompatible version of the Instance-ID SDK in parallel can break Firebase Targeting, as the same App-Instance will get 2 different identifiers. Ensuring that we do not break Firebase Targeting had the highest priority when discussing the design of FIS SDKs. There is a rather short summary in go/fis-iid-incompatible-versions.
In that sense, don't you think it is desirable to prevent this situation with a breaking (checked) exception?

vkryachko replied:
I guess one could argue either way, but imo the severity of the condition that causes an exception to be thrown is orthogonal to the type of exception to throw(`checked/unchecked`).

To quote effective java(go/ej3e-70):
```
 use checked exceptions for conditions from which the caller
can reasonably be expected to recover.
...By confronting the user with a checked exception, the API designer presents a
mandate to recover from the condition
```

With this in mind, it does not sound like something that can be recovered from at runtime, and the only way to recover is to rebuild the app with a different version of iid, hence my suggestion to make it unchecked. But if there are reasonable ways to recover from this state then sure, it could be checked.

ankitaj224 replied:
@vkryachko Accepted your suggestion and updated the code. PTAL.

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -133,11 +147,23 @@ public static FirebaseInstallations getInstance() {
+  public static FirebaseInstallations getInstance(@NonNull FirebaseApp app)
+      throws FirebaseInstallationsException {
     Preconditions.checkArgument(app != null, "Null is not a valid value of FirebaseApp.");
+    preventParallelUsageOfIncompatibleVersionOfInstanceIdSDk();
```
+1
I think, we only need to check once in the private constructor of FirebaseInstallations.
(and maybe we only need to check for one FirebaseApp, e.g. the [default] app)

vkryachko replied:
On second thought, I think additionally you may want to move the check into the constructor of `FirebaseInstallations`, not the getInstance() method.
Motivation:
* Any other SDK that depends on FIS interop would respect the check automatically, as component dependency resolution does not normally go through `getInstance()` methods

vkryachko replied:
re my last comment: seems like this is exactly what Rayo suggested above :)

ankitaj224 replied:
@vkryachko Moved the check to the constructor. Do you think we still need to make it async?

vkryachko commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -72,6 +78,14 @@
+  private static final String IID_SDK_PROP_FILE = "/firebase-iid.properties";
+  /* List of all Instance-ID versions that are incompatible with FIS SDK.*/
+  private static final List<String> IID_SDK_VERSIONS_INCOMPATIBLE_WITH_FIS =
+      Arrays.asList("20.0.2", "20.0.1", "20.0.0", "19.0.1", "19.0.0", "18.0.0");
```
What if I use an even older version? like 17.1.2?

ankitaj224 replied:
Changed it to check for any older version than 20.1.1 ( IID with FIS)

ankitaj224 commented:
@vkryachko  and @andirayo 

Updated this PR to handle runtime check for incompatible IID versions. Addressed the older comments. Please review the changes again.

I plan on implementing the lint checks as suggested by vlad in another PR.

Thanks.

vkryachko commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -124,6 +134,51 @@ public Thread newThread(Runnable r) {
     this.utils = utils;
     this.iidStore = iidStore;
     this.fidGenerator = fidGenerator;
+    preventParallelUsageOfIncompatibleVersionOfInstanceIdSDk();
```
Why do it in a blocking way? The least we can do is do it on a thread and post the exception to the caller thread(or UI thread) through its Looper

vkryachko commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -124,6 +134,51 @@ public Thread newThread(Runnable r) {
+    String iidVersion = getIIDVersion();
+    // IID SDK version is incompatible if it is older than the IID SDK version with FIS
+    // dependency(i.e 20.1.1).
+    if (TextUtils.isEmpty(iidVersion)
```
afaict this check is not future-proof as iid won't have a properties file once released from github(something @VinayGuthal is working on)

fyi @ashwinraghav

ankitaj224 replied:
I wasn't aware of this. Clients moving from IID to directly depend on FIS have been requesting such a check to avoid running into multiple firebase identifiers. Without such a check in place we cant be sure that there is only one identifier.

andirayo replied:
Question:
Why are we making changes to FirebaseInstanceId, when the plan is to deprecate this SDK as soon as possible?
Can't we just leave FirebaseInstanceId the way it is right now and focus on moving all Firebase SDKs to directly depend on FIS?

vkryachko commented:
> I plan on implementing the lint checks as suggested by vlad in another PR.

With this check planned, do you think it may be sufficient so that the runtime check is not needed? (as lint will trigger on all gradle builds making runtime failure very unlikely to happen)

ankitaj224 commented:
> > I plan on implementing the lint checks as suggested by vlad in another PR.
> 
> With this check planned, do you think it may be sufficient so that the runtime check is not needed? (as lint will trigger on all gradle builds making runtime failure very unlikely to happen)

The lint check definitely precedes over the runtime check. However, I am trying to understand how to implement the lint check which will take sometime on my end. Meanwhile, the clients migrating to FIS had the concern about no incompatibility check at all. The runtime check was already in place will unblock the clients, I thought of moving forward with this PR before the lint check.

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -27,6 +28,7 @@
 import com.google.firebase.FirebaseException;
 import com.google.firebase.FirebaseOptions;
 import com.google.firebase.heartbeatinfo.HeartBeatInfo;
+import com.google.firebase.iid.FirebaseInstanceId;
```
How can this work if there is no Instance ID available?

ankitaj224 replied:
Check the isIIDVersionCompatible(). If there is no Instance ID it catches NoClassDefFoundError and returns compatibility true.

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -76,6 +80,12 @@
+  /* File used to read the version of Instance-ID(IID) SDK if it exists in the dependency tree.
+  This is needed to prevent incompatible versions of IID SDK to be used in parallel with FIS SDK. */
+  private static final String IID_SDK_PROP_FILE = "/firebase-iid.properties";
+  private static final String IID_VERSION_WITH_FIS = "20.1.1";
```
Actually IID v20.1.0 already depended on FIS, but I assume we want to make sure they use the hotfixed version? right?

If that is the case, please add an appropriate comment, example:

// FirebaseInstanceId v20.1.0 already depended on FirebaseInstallations for identifier creation, but had an issue of possible silent failures. Thus, v20.1.1 or later is recommended.

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -124,6 +134,51 @@ public Thread newThread(Runnable r) {
+    if (!isIIDVersionCompatible()) {
+      throw new IllegalStateException(
+          "FirebaseInstallations will not work correctly with current version of Firebase "
+              + "Instance ID SDK. Please update your Firebase Instance ID version.");
```
These version of FirebaseInstanceId and FirebaseInstallations can not be used in parallel.
Please update your Firebase Instance ID to a version 20.1.0 or later.

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -124,6 +134,51 @@ public Thread newThread(Runnable r) {
+   */
+  private boolean isIIDVersionCompatible() {
+    try {
+      final Class<FirebaseInstanceId> firebaseInstanceIdClass = FirebaseInstanceId.class;
```
How about using the fully qualified class path here?
com.google.firebase.iid.FirebaseInstanceId.class

andirayo commented on firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:
```diff
@@ -124,6 +134,51 @@ public Thread newThread(Runnable r) {
+        && utils.compareVersions(IID_VERSION_WITH_FIS, iidVersion) >= 1) {
+      return false;
+    }
+    return true;
```
if (TextUtils.isEmpty(iidVersion)) {
  return false;
}

return 1 > utils.compareVersions(IID_VERSION_WITH_FIS, iidVersion);

vkryachko commented:
fyi I've prototyped a lint check in #1346 , ptal
